sdk/include/platform/ibex/platform-hardware_revoker.hh

```diff
@@ -161,6 +192,43 @@ namespace Ibex
 			// value because we know it will be the last value + 1.
 			epoch++;
 		}
+
+		/**
+		 * Block until the revocation epoch specified by `epoch` has completed.
+		 */
+		bool wait_for_completion(Timeout *timeout, uint32_t epoch)
+		{
+			uint32_t interruptValue;
+			do
+			{
+				// Read the current interrupt futex word.  We want to retry if
+				// an interrupt happens after this point.
+				interruptValue = *interruptFutex;
+				// Make sure that the compiler doesn't reorder the read of the
+				// futex word with respect to the read of the revocation epoch.
+				__c11_atomic_signal_fence(__ATOMIC_SEQ_CST);
+				// If the requested epoch has finished, return success.
+				if (has_revocation_finished_for_epoch<true>(epoch))
+				{
+					return true;
+				}
+				// Request the interrupt
+				revoker_device().interruptRequested = 1;
+				// There is a possible race: if the revocation pass finished
+				// before we requested the interrupt, we won't get the
+				// interrupt.  Check again before we wait.
+				if (has_revocation_finished_for_epoch<true>(epoch))
```

Comment:
Something about this seems fishy.  If it completed after our read of `*interruptFutex` above, then `futex_timed_wait` should bounce us out.  So for this asynchrony to result in us missing a wakeup, it would have had to complete before that read, but we checked `has_revocation_finished` after that read.

Reply:
No, the problem is that an interrupt for sweep n doesn’t request an interrupt, one is not delivered. If we check for completion, find the epoch is not done, then it finished, then we request the interrupt. The request for the interrupt is dropped (no revocation sweep ongoing, but I’ve if it isn’t then the interrupt won’t retroactively fire for the last epoch), so the futex word isn’t updated. We then wait until some other thread enters this code and requests an interrupt, which actually does fire (or forever, or until the timeout).

Reply:
Oh, I see.  Is the identical check on line 211 just a micro-optimization to avoid requesting an interrupt if the epoch under test is very far in the past?

Reply:
Yes, it would cause an interrupt to be raised that no one is looking for, so we’d transition into the scheduler and back for no reason, which costs a few hundred cycles, whereas this check should cost less than 10.